src/api_gateway/handlers/customer_handler/customer_handler.go

```diff
@@ -2,24 +2,170 @@ package customer_handler
 
 import (
 	"log"
+	"monorepo/src/api_gateway/utils"
+	"monorepo/src/libs/etc"
+	"monorepo/src/libs/redis"
+	libs "monorepo/src/libs/utils"
 	"net/http"
+	"time"
 
 	"github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/ext"
 )
 
 type CustomerHandlers interface {
 	TestHandler2(w http.ResponseWriter, r *http.Request)
+	VerfyCodeHandler(w http.ResponseWriter, r *http.Request)
+	SendCodeHandler(w http.ResponseWriter, r *http.Request)
 }
 
 type customerHandler struct {
-	tracer opentracing.Tracer
+	tracer  opentracing.Tracer
+	redisDB redis.InMemoryStorageI
 }
 
 // New creates handler
 func New(logger *log.Logger, tracer opentracing.Tracer) CustomerHandlers {
-	return &customerHandler{tracer: tracer}
+	return &customerHandler{
+		tracer: tracer,
+		// redisDB: rds,
+	}
 }
 func (ch *customerHandler) TestHandler2(w http.ResponseWriter, r *http.Request) {
 	// ch.logger.Print("Got a request.")
 	w.Write([]byte("Hello, World2!"))
 }
+
+func (ch *customerHandler) SendCodeHandler(w http.ResponseWriter, r *http.Request) {
+	sendCodeHandlerSpan := ch.tracer.StartSpan("SendCodeHandler")
+	defer sendCodeHandlerSpan.Finish()
+
+	// Set some tags on the clientSpan to annotate that it's the client span. The additional HTTP tags are useful for debugging purposes.
+	ext.SpanKindRPCClient.Set(sendCodeHandlerSpan)
+	ext.HTTPUrl.Set(sendCodeHandlerSpan, r.URL.Path)
+	ext.HTTPMethod.Set(sendCodeHandlerSpan, "GET")
+
+	// Inject the client span context into the headers
+	//3
+	ch.tracer.Inject(sendCodeHandlerSpan.Context(), opentracing.HTTPHeaders, opentracing.HTTPHeadersCarrier(r.Header))
+	time.Sleep(2 * time.Second)
+
+	// ctx:=opentracing.ContextWithSpan(r.Context(), testHandlerSpan)
+	// send this ctx to services called here
+
+	// read body from request
+	var body utils.ReqSendCode
+	if err := utils.BodyParser(r, body); err != nil {
+		utils.HandleBadRequestErrWithMessage(w, err, "invalid body")
+		return
+	}
+
+	// generate one-time-password
+	code := etc.GenerateCode(4, true)
+
+	// check if request was sent by phone number
+	if body.Phone != "" {
+		// send code to the number
+		if err := libs.SendCodeToPhone(body.Phone, code); err != nil {
+			utils.HandleBadRequestErrWithMessage(w, err, "invalid phone number")
```

Comment:
validate phoneNumber before sending SMS. In general validate request fields, just after parsing body.

Reply:
create isPhoneNumber, isEmail  validators using regex in libs package. And use this to verify the body fields. Make the login endpoint receiving field name universal for both email and phoneNumber (maybe loginValue).